### PR TITLE
Use systemd cgroup driver for v1.24.0+

### PR DESCRIPTION
**Pull request by BenTheElder**
see: #1726

### CODE REVIEW

File: images/base/files/etc/containerd/config.toml [hunk 1]
```diff
@@ -19,6 +19,9 @@ version = 2
   runtime_type = "io.containerd.runc.v2"
   # Generated by "ctr oci spec" and modified at base container to mount poduct_uuid
   base_runtime_spec = "/etc/containerd/cri-base.json"
+  [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
+    # use systemd cgroup by default
+    SystemdCgroup = true
```
Comment (BenTheElder) [inline images/base/files/etc/containerd/config.toml:24]: may also need to do this for test-handler.

File: pkg/build/nodeimage/containerd.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+	"sigs.k8s.io/kind/pkg/errors"
+	"sigs.k8s.io/kind/pkg/exec"
+
+	"sigs.k8s.io/kind/pkg/internal/patch"
+)
+
+const containerdConfigPath = "/etc/containerd/config.toml"
+
+const containerdConfigPatchSystemdCgroupFalse = `
```
Comment (BenTheElder) [inline pkg/build/nodeimage/containerd.go:30]: should also patch test-handler here

Comment (BenTheElder) [reply pkg/build/nodeimage/containerd.go:30]: done.

File: images/base/files/etc/containerd/config.toml [hunk 3]
```diff
@@ -19,11 +19,18 @@ version = 2
...
 
 # Setup a runtime with the magic name ("test-handler") used for Kubernetes
 # runtime class tests ...
 [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.test-handler]
+  # same settings as runc
```
Comment (aojea) [inline images/base/files/etc/containerd/config.toml:29]: What is this test-handler used for?

Comment (BenTheElder) [reply images/base/files/etc/containerd/config.toml:29]: see comment two lines above in file? :-)

Comment (BenTheElder) [reply images/base/files/etc/containerd/config.toml:29]: https://github.com/kubernetes/kubernetes/blob/e045e5c875b2f51c4c4b1bc14d4e717123649b17/test/e2e/framework/node/runtimeclass.go

it's a tricky problem in k/k e2e testing. to test that runtimeclass works, but need to know that a non-default runtime is configured.

**Comment by BenTheElder** [comment]
hmm 1.24 kubelet unhealthy 
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/2737/pull-kind-e2e-kubernetes/1521573229309726720

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/2737/pull-kind-e2e-kubernetes/1521573229309726720/artifacts/logs/kind-control-plane/kubelet.log

> May 03 19:52:45 kind-control-plane kubelet[1749]: Error: failed to run Kubelet: invalid configuration: cgroup ["kubelet"] has some missing paths: /sys/fs/cgroup/pids/kubelet.slice, /sys/fs/cgroup/cpu,cpuacct/kubelet.slice, /sys/fs/cgroup/cpu,cpuacct/kubelet.slice, /sys/fs/cgroup/cpuset/kubelet.slice, /sys/fs/cgroup/memory/kubelet.slice, /sys/fs/cgroup/systemd/kubelet.slice, /sys/fs/cgroup/hugetlb/kubelet.slice

**Comment by BenTheElder** [comment]
right ... all the entrypoint tricks are aimed at cgroupfs.

File: images/base/files/etc/containerd/config.toml (hunk 3 above)
Comment (aojea) [reply images/base/files/etc/containerd/config.toml:29]: I was wondering why do we need to fill this too, but seems that is for coherence,

Comment (aojea) [reply images/base/files/etc/containerd/config.toml:29]: found it in the 5th commit

**Comment by BenTheElder** [comment]
https://github.com/opencontainers/runc/blob/main/docs/systemd.md

**Comment by BenTheElder** [comment]
Got back to this a bit today:
- kubelet appends .slice internally and you should not add it yourself
- ... but the cgroup `--cgroup-root` + .slice must exist
- if we use the current `/kubelet` cgroup root approach but populate `/kubelet.slice` instead, everything seems fine

Debating if we should plumb through a switch or just always setup both, leaning towards the latter so it's simpler to ensure either can be used.

**Comment by BenTheElder** [comment]
s390x build flake https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/2737/pull-kind-build/1522383838079094784

File: images/base/files/etc/containerd/config.toml (hunk 3 above)
Comment (BenTheElder) [reply images/base/files/etc/containerd/config.toml:29]: in e2e we will actually create pod(s) with this runtime, we need that to match kubelet wrt cgroup driver

**Comment by BenTheElder** [comment]
s390x flake again
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/2737/pull-kind-build/1522405658077433856
/retest

**Comment by BenTheElder** [comment]
logs still show failing to remove rmda cgroup but that's not new to this PR / driver https://github.com/kubernetes/kubernetes/issues/109182

**Comment by BenTheElder** [comment]
We need to test this in more environments before moving forward, just to be on the safe side that we haven't missed a quirk.

I'm pushing a node image from this, will update the PR with it so that github actions environments pick it up, and then see about getting it tested in some additional environments.

**Comment by BenTheElder** [comment]
podman the node only got this far:
```
INFO: ensuring we can execute mount/umount even with userns-remap
INFO: remounting /sys read-only
INFO: making mounts shared
INFO: detected cgroup v1
WARN: cgroupns not enabled! Please use cgroup v2, or cgroup v1 with cgroupns enabled.
INFO: fix cgroup mounts for all subsystems
```

exit status 1 https://github.com/kubernetes-sigs/kind/actions/runs/2296780194

**Comment by BenTheElder** [comment]
Across the board broken in podman, not in docker. https://github.com/kubernetes-sigs/kind/pull/2737/checks

/hold

**Comment by BenTheElder** [comment]
Also cgroup v2 across the board in CI https://github.com/kubernetes-sigs/kind/runs/6359585961?check_suite_focus=true
... but not locally on my machine 🤔

**Comment by BenTheElder** [comment]
ah, cgroupv2 failed on setting up the vagrant env.

**Comment by BenTheElder** [comment]
Figured as much, containerd has the same issue. https://github.com/containerd/containerd/pull/6915

https://github.blog/2022-04-12-git-security-vulnerability-announced/

**Comment by BenTheElder** [comment]
OK, so it's now just (podman, cgroup v1) that is broken, all of cgroup v2 is passing after #2742

**Comment by BenTheElder** [comment]
trace from https://github.com/kubernetes-sigs/kind/pull/2737/commits/9e14207e657296ca10946745ca024ffdd212b72c
```
+ source /usr/local/bin/entrypoint
++ set -o errexit
++ set -o nounset
++ set -o pipefail
++ userns=
++ grep -Eqv '0[[:space:]]+0[[:space:]]+4294967295' /proc/self/uid_map
++ validate_userns
++ [[ -z '' ]]
++ return
++ configure_containerd
++ local snapshotter=
++ [[ -n '' ]]
++ [[ -z '' ]]
+++ stat -f -c %T /kind
++ [[ overlayfs == \z\f\s ]]
++ [[ -n '' ]]
++ configure_proxy
++ mkdir -p /etc/systemd/system.conf.d/
++ cat
++ fix_kmsg
++ [[ ! -e /dev/kmsg ]]
++ fix_mount
++ echo 'INFO: ensuring we can execute mount/umount even with userns-remap'
INFO: ensuring we can execute mount/umount even with userns-remap
+++ which mount
+++ which umount
++ chown root:root /usr/bin/mount /usr/bin/umount
+++ which mount
+++ which umount
++ chmod -s /usr/bin/mount /usr/bin/umount
++++ which mount
+++ stat -f -c %T /usr/bin/mount
++ [[ overlayfs == \a\u\f\s ]]
++ echo 'INFO: remounting /sys read-only'
INFO: remounting /sys read-only
++ mount -o remount,ro /sys
++ echo 'INFO: making mounts shared'
INFO: making mounts shared
++ mount --make-rshared /
++ fix_cgroup
++ [[ -f /sys/fs/cgroup/cgroup.controllers ]]
++ echo 'INFO: detected cgroup v1'
INFO: detected cgroup v1
++ local current_cgroup
+++ grep -E '^[^:]*:([^:]*,)?cpu(,[^,:]*)?:.*' /proc/self/cgroup
+++ cut -d: -f3
++ current_cgroup=/machine.slice/libpod-956ff18413c2d67f2a684aab03727625784586aa3398f2b03da84763df5155e8.scope
++ '[' /machine.slice/libpod-956ff18413c2d67f2a684aab03727625784586aa3398f2b03da84763df5155e8.scope = / ']'
++ echo 'WARN: cgroupns not enabled! Please use cgroup v2, or cgroup v1 with cgroupns enabled.'
WARN: cgroupns not enabled! Please use cgroup v2, or cgroup v1 with cgroupns enabled.
++ echo 'INFO: fix cgroup mounts for all subsystems'
INFO: fix cgroup mounts for all subsystems
++ local cgroup_subsystems
+++ findmnt -lun -o source,target -t cgroup
+++ awk '{print $2}'
+++ grep /machine.slice/libpod-956ff18413c2d67f2a684aab03727625784586aa3398f2b03da84763df5155e8.scope
++ cgroup_subsystems='/sys/fs/cgroup/blkio
/sys/fs/cgroup/memory
/sys/fs/cgroup/cpu,cpuacct
/sys/fs/cgroup/rdma
/sys/fs/cgroup/freezer
/sys/fs/cgroup/cpuset
/sys/fs/cgroup/perf_event
/sys/fs/cgroup/pids
/sys/fs/cgroup/devices
/sys/fs/cgroup/misc
/sys/fs/cgroup/hugetlb
/sys/fs/cgroup/net_cls,net_prio
/sys/fs/cgroup/systemd'
++ local unsupported_cgroups
+++ grep -v /machine.slice/libpod-956ff18413c2d67f2a684aab03727625784586aa3398f2b03da84763df5155e8.scope
+++ findmnt -lun -o source,target -t cgroup
+++ awk '{print $2}'
++ unsupported_cgroups=
```

**Comment by BenTheElder** [comment]
So this is just podman runtime being updated and  `unsupported_cgroups=$(findmnt -lun -o source,target -t cgroup | grep -v "${current_cgroup}" | awk '{print $2}')` failing due to their being no match.

That's not related to this PR. It's a minor bug in https://github.com/kubernetes-sigs/kind/pull/2709 which hasn't been picked up yet due to not previously bumping the node image + podman CI uses default node image.

I'll send a fix for that, then we can rebase this PR and try again.

**Comment by BenTheElder** [comment]
rebasing on fix in https://github.com/kubernetes-sigs/kind/pull/2743

**Comment by BenTheElder** [comment]
And now podman is passing https://github.com/kubernetes-sigs/kind/runs/6362603916?check_suite_focus=true

**Comment by BenTheElder** [comment]
> May 10 02:34:53.338: INFO: externalsvc-72bcl[services-5001].container[externalsvc]=failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: unable to apply cgroup configuration: failed to write 84042: write /sys/fs/cgroup/rdma/kubelet.slice/kubelet-kubepods.slice/kubelet-kubepods-besteffort.slice/kubelet-kubepods-besteffort-podbd0b7b96_aa32_4e00_8323_6f20dd1b85a4.slice/cri-containerd-6e1a4e9c43aba1945e9a28e410939e0801c180fa69bb1850018c9c0b2956a91a.scope/cgroup.procs: no such device: unknown

we still have the rdma issue https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kind/2737/pull-kind-conformance-parallel-ipv6/1523846989832261632

but that's not new
/retest

File: pkg/cluster/internal/kubeadm/config.go [hunk 4]
```diff
@@ -621,6 +627,13 @@ func Config(data ConfigData) (config string, err error) {
 		data.KubeadmFeatureGates["IPv6DualStack"] = true
 	}
 
+	// before 1.24 kind uses cgroupfs
+	// after 1.24 kind uses systemd starting in kind v0.13.0
+	// before kind v0.13.0 kubernetes 1.24 wasn't released yet
+	if ver.LessThan(version.MustParseSemantic("v1.24.0")) {
+		data.CgroupDriver = "cgroupfs"
```
Comment (aojea) [inline pkg/cluster/internal/kubeadm/config.go:634]: this effectively means that  `ConfigTemplateBetaV1` and `ConfigTemplateBetaV2` are going to use data.CgroupDriver = "cgroupfs", maybe for consistency is better to replace in all templates?

File: images/base/files/usr/local/bin/entrypoint [hunk 5]
```diff
@@ -294,10 +294,12 @@ fix_cgroup() {
   # "nesting" clusters, unless we instruct it to use a different cgroup root.
   # We do this, and when doing so we must fixup this alternative root
   # currently this is hardcoded to be /kubelet
+  # under systemd cgroup driver, kubelet appends .slice
```
Comment (aojea) [inline images/base/files/usr/local/bin/entrypoint:297]: is there any problem for mounting it twice?
is not possible to detect which one we want?

Comment (BenTheElder) [reply images/base/files/usr/local/bin/entrypoint:297]: No there's no problem. The cgroup controllers are enabled on many trees. Only one of these two will be used at runtime.

Doing it this way is robust in the face of future changes and user config patches.
We don't know what the node config will be until after we spawn the node and discover details like Kubernetes version.

File: pkg/cluster/internal/kubeadm/config.go (hunk 4 above)
Comment (BenTheElder) [reply pkg/cluster/internal/kubeadm/config.go:634]: Do you mean don't template it?

I don't think we should do that no, selecting based on Kubernetes version is the current approach and this is more explicit about how that is selected. Templating it everywhere ensures greater logical consistency.

File: pkg/build/nodeimage/containerd.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
+  SystemdCgroup = false
+
+[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.test-handler.options]
+  SystemdCgroup = false
+`
+
+func configureContainerdSystemdCgroupFalse(containerCmdr exec.Cmder, config string) error {
```
Comment (aojea) [inline pkg/build/nodeimage/containerd.go:38]: can this fail against an image that was built for 1.24.x with an old version of kind, without the containerd change?

Comment (BenTheElder) [reply pkg/build/nodeimage/containerd.go:38]: yes, users that upgrade to kind v0.13 will need to use an image built with a newer version
we have no official images that way.

for custom built images that are 1.24+ with kind < v0.13, users will need to build new images. we'll have a release note.

the opposite is also true, 1.24+ official images / images built going forward will require v0.13.

this already would have happened in 1.21 with kubeadm defaulting to systemd then, we just shifted it a few releases.

**Comment by BenTheElder** [comment]
Working on v0.13.0 based on this, pushing images. Wrote release notes.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes-sigs/kind/releases/tag/v0.13.0

**Pull request merged**